com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/ProtocolServer.java

```diff
@@ -75,43 +80,83 @@ public void sendResponse(Messages.Response response) {
+            synchronized (this) {
+                this.isDispatchingRequest = true;
             }
 
-            if (ex instanceof VMDisconnectedException) {
-                // mark it success to avoid reporting error on VSCode.
-                response.success = true;
-                sendResponse(response);
-            } else {
-                String exceptionMessage = ex.getMessage() != null ? ex.getMessage() : ex.toString();
-                ErrorCode errorCode = ex instanceof DebugException ? ErrorCode.parse(((DebugException) ex).getErrorCode()) : ErrorCode.UNKNOWN_FAILURE;
-                boolean isUserError = ex instanceof DebugException && ((DebugException) ex).isUserError();
-                if (isUserError) {
-                    usageDataSession.recordUserError(errorCode);
+            debugAdapter.dispatchRequest(request).thenCompose((response) -> {
+                CompletableFuture<Void> future = new CompletableFuture<>();
+                if (response != null) {
+                    sendResponse(response);
+                    future.complete(null);
                 } else {
-                    logger.log(Level.SEVERE, String.format("[error response][%s]: %s", request.command, exceptionMessage), ex);
+                    future.completeExceptionally(new DebugException("The request dispatcher should not return null response.",
+                            ErrorCode.UNKNOWN_FAILURE.getId()));
                 }
+                return future;
+            }).exceptionally((ex) -> {
+                Messages.Response response = new Messages.Response(request.seq, request.command);
+                if (ex instanceof CompletionException && ex.getCause() != null) {
+                    ex = ex.getCause();
+                }
+
+                if (ex instanceof VMDisconnectedException) {
+                    // mark it success to avoid reporting error on VSCode.
+                    response.success = true;
+                    sendResponse(response);
+                } else {
+                    String exceptionMessage = ex.getMessage() != null ? ex.getMessage() : ex.toString();
+                    ErrorCode errorCode = ex instanceof DebugException ? ErrorCode.parse(((DebugException) ex).getErrorCode()) : ErrorCode.UNKNOWN_FAILURE;
+                    boolean isUserError = ex instanceof DebugException && ((DebugException) ex).isUserError();
+                    if (isUserError) {
+                        usageDataSession.recordUserError(errorCode);
+                    } else {
+                        logger.log(Level.SEVERE, String.format("[error response][%s]: %s", request.command, exceptionMessage), ex);
+                    }
+
+                    sendResponse(AdapterUtils.setErrorResponse(response,
+                            errorCode,
+                            exceptionMessage));
+                }
+                return null;
+            }).join();
+        } finally {
+            synchronized (this) {
```

Comment:
Avoid lock the whole object, use private object for lock.

Reply:
fixed.